Exception in thread "main" brut.androlib.AndrolibException: Could not decode arsc file

```
Hamids-MBP:apk Hamid$ ./apktool d roqya.apk
I: Using Apktool 2.0.0 on roqya.apk
I: Loading resource table...
Exception in thread "main" brut.androlib.AndrolibException: Could not decode arsc file
	at brut.androlib.res.decoder.ARSCDecoder.decode(ARSCDecoder.java:52)
	at brut.androlib.res.AndrolibResources.getResPackagesFromApk(AndrolibResources.java:538)
	at brut.androlib.res.AndrolibResources.loadMainPkg(AndrolibResources.java:63)
	at brut.androlib.res.AndrolibResources.getResTable(AndrolibResources.java:55)
	at brut.androlib.Androlib.getResTable(Androlib.java:64)
	at brut.androlib.ApkDecoder.setTargetSdkVersion(ApkDecoder.java:209)
	at brut.androlib.ApkDecoder.decode(ApkDecoder.java:92)
	at brut.apktool.Main.cmdDecode(Main.java:165)
	at brut.apktool.Main.main(Main.java:81)
Caused by: java.io.IOException: Expected: 0x00000008, got: 0x00000001
	at brut.util.ExtDataInput.skipCheckShort(ExtDataInput.java:56)
	at brut.androlib.res.decoder.ARSCDecoder.readValue(ARSCDecoder.java:238)
	at brut.androlib.res.decoder.ARSCDecoder.readEntry(ARSCDecoder.java:201)
	at brut.androlib.res.decoder.ARSCDecoder.readConfig(ARSCDecoder.java:189)
	at brut.androlib.res.decoder.ARSCDecoder.readType(ARSCDecoder.java:157)
	at brut.androlib.res.decoder.ARSCDecoder.readPackage(ARSCDecoder.java:114)
	at brut.androlib.res.decoder.ARSCDecoder.readTable(ARSCDecoder.java:78)
	at brut.androlib.res.decoder.ARSCDecoder.decode(ARSCDecoder.java:47)
	... 8 more
```

I'm getting the same issue on the new TW M builds with the latest apktool release and master branch:
> I: Using Apktool 2.1.0-a64a03-SNAPSHOT on framework-res.apk
> I: Loading resource table...
> Exception in thread "main" brut.androlib.AndrolibException: Could not decode arsc file
>         at brut.androlib.res.decoder.ARSCDecoder.decode(ARSCDecoder.java:52)
>         at brut.androlib.res.AndrolibResources.getResPackagesFromApk(AndrolibResources.java:544)
>         at brut.androlib.res.AndrolibResources.loadMainPkg(AndrolibResources.java:63)
>         at brut.androlib.res.AndrolibResources.getResTable(AndrolibResources.java:55)
>         at brut.androlib.Androlib.getResTable(Androlib.java:67)
>         at brut.androlib.ApkDecoder.setTargetSdkVersion(ApkDecoder.java:198)
>         at brut.androlib.ApkDecoder.decode(ApkDecoder.java:96)
>         at brut.apktool.Main.cmdDecode(Main.java:165)
>         at brut.apktool.Main.main(Main.java:81)
> Caused by: java.io.IOException: Expected: 0x00000008, got: 0x00000000
>         at brut.util.ExtDataInput.skipCheckShort(ExtDataInput.java:56)
>         at brut.androlib.res.decoder.ARSCDecoder.readValue(ARSCDecoder.java:282)
>         at brut.androlib.res.decoder.ARSCDecoder.readEntry(ARSCDecoder.java:224)
>         at brut.androlib.res.decoder.ARSCDecoder.readTableType(ARSCDecoder.java:212)
>         at brut.androlib.res.decoder.ARSCDecoder.readTableTypeSpec(ARSCDecoder.java:154)
>         at brut.androlib.res.decoder.ARSCDecoder.readTablePackage(ARSCDecoder.java:116)
>         at brut.androlib.res.decoder.ARSCDecoder.readTableHeader(ARSCDecoder.java:78)
>         at brut.androlib.res.decoder.ARSCDecoder.decode(ARSCDecoder.java:47)
>         ... 8 more
 
Apk: https://www.dropbox.com/s/i80xi2uyjlfttji/framework-res.apk?dl=0

--- a/brut.apktool/apktool-lib/src/main/java/brut/androlib/res/decoder/ARSCDecoder.java
+++ b/brut.apktool/apktool-lib/src/main/java/brut/androlib/res/decoder/ARSCDecoder.java
@@ -54,11 +54,10 @@ public class ARSCDecoder {
     }
 
     private ARSCDecoder(InputStream arscStream, ResTable resTable, boolean storeFlagsOffsets, boolean keepBroken) {
+        arscStream = mCountIn = new CountingInputStream(arscStream);
         if (storeFlagsOffsets) {
-            arscStream = mCountIn = new CountingInputStream(arscStream);
             mFlagsOffsets = new ArrayList<FlagsOffset>();
         } else {
-            mCountIn = null;
             mFlagsOffsets = null;
         }
         mIn = new ExtDataInput(new LittleEndianDataInputStream(arscStream));
@@ -186,12 +185,18 @@ public class ARSCDecoder {
 
         /* res0, res1 */mIn.skipBytes(3);
         int entryCount = mIn.readInt();
-        /* entriesStart */mIn.skipInt();
-
+        int entriesStart = mIn.readInt();
         mMissingResSpecs = new boolean[entryCount];
         Arrays.fill(mMissingResSpecs, true);
 
         ResConfigFlags flags = readConfigFlags();
+        int position = (mHeader.startPosition + entriesStart) - (entryCount * 4);
+
+        // For some APKs there is a disconnect between the reported size of Configs
+        // If we find a mismatch skip those bytes.
+        if (position != mCountIn.getCount()) {
+            mIn.skipBytes(position - mCountIn.getCount());
+        }
         int[] entryOffsets = mIn.readIntArray(entryCount);
 
         if (flags.isInvalid) {
@@ -217,7 +222,10 @@ public class ARSCDecoder {
     }
 
     private void readEntry() throws IOException, AndrolibException {
-		/* size */mIn.skipBytes(2);
+        short size = mIn.readShort();
+        if (size < 0) {
+            throw new AndrolibException("Entry size is under 0 bytes.");
+        }
         short flags = mIn.readShort();
         int specNamesId = mIn.readInt();
 
@@ -464,7 +472,7 @@ public class ARSCDecoder {
     }
 
     private Header nextChunk() throws IOException {
-        return mHeader = Header.read(mIn);
+        return mHeader = Header.read(mIn, mCountIn);
     }
 
     private void checkChunkType(int expectedType) throws AndrolibException {
@@ -502,21 +510,26 @@ public class ARSCDecoder {
         public final short type;
         public final int headerSize;
         public final int chunkSize;
+        public final int startPosition;
+        public final int endPosition;
 
-        public Header(short type, int headerSize, int chunkSize) {
+        public Header(short type, int headerSize, int chunkSize, int headerStart) {
             this.type = type;
             this.headerSize = headerSize;
             this.chunkSize = chunkSize;
+            this.startPosition = headerStart;
+            this.endPosition = headerStart + chunkSize;
         }
 
-        public static Header read(ExtDataInput in) throws IOException {
+        public static Header read(ExtDataInput in, CountingInputStream countIn) throws IOException {
             short type;
+            int start = countIn.getCount();
             try {
                 type = in.readShort();
             } catch (EOFException ex) {
-                return new Header(TYPE_NONE, 0, 0);
+                return new Header(TYPE_NONE, 0, 0, countIn.getCount());
             }
-            return new Header(type, in.readShort(), in.readInt());
+            return new Header(type, in.readShort(), in.readInt(), start);
         }
 
         public final static short TYPE_NONE = -1, TYPE_TABLE = 0x0002,

